graphlearn.py

```diff
@@ -171,137 +217,185 @@ def sample_set(self, graph_iter, improvement_rules=
-                md[k2]=v2
-            shelve[k] = md
-        self.grammar_functions.__dict__.pop('data')
-        self.substitute_grammar = shelve
+        self.same_radius= same_radius
+        self.similarity=similarity
+        self.sampling_interval=sampling_interval
+        self.n_steps=n_steps
+        self.n_jobs=n_jobs
+
+        # multiprocessing work-around oOoOoo sadly necessary
+        self.postprocessing=dill.dumps(postprocessing)
+
+        # adapt grammar to task:
+        self.grammar_preprocessing()
 
         # do the improvement
         if n_jobs == 0:
             for graph in graph_iter:
-                yield self.sample(graph)
+                yield self._sample(graph)
         else:
+            # make it so that we dont need to copy the whole grammar a few times  for multiprocessing
             problems = itertools.izip(
                 graph_iter, itertools.repeat(self))
             pool = Pool(processes=n_jobs)
-            batch_size=self.improvement_rules['batch_size']
+            batch_size=batch_size
             it = pool.imap_unordered(improve_loop_multi, problems, batch_size)
-            for liste in it:
-                yield liste
+            for pair in it:
+                yield pair
+            pool.close()
 
 
-    def sample(self, graph):
-        # prepare variables and graph
 
-        graph = myutils.expand_edges(graph)
+
+    def similarity_checker(self,graph,set_reference=False):
+        '''
+        always check if similarity is relevant.. if so then:
+
+        if set_reference is True:
+            remember the vectorized object
+        else:
+            similarity between start graph and current graph is expected to decrease.
+            if similarity meassure smaller than the limit, we stop
+            because we dont want to drift further
+        '''
+        if self.similarity>0:
+            if set_reference:
+                self.vectorizer_expanded._reference_vec = \
+                    self.vectorizer_expanded._convert_dict_to_sparse_matrix(
+                    self.vectorizer_expanded._transform(0,nx.Graph(graph)) )
+            else:
+                similarity = self.vectorizer_expanded._similarity(graph,[1])
+                return similarity < self.similarity
+        return False
+
+
+
+
+    def _sample_init(self,graph):
+        '''
+        we prepare a single graph to be sampled
+
+        - first we expand its edges to nodes, so eden will be able wo work its magic on it
+        - then we calculate a score for the graph, to see how much we like it
+        - in the end we set up one of the stop conditions for the sample loop
+        '''
+        graph = graphlearn_utils.expand_edges(graph)
         self.score(graph)
+        self.similarity_checker(graph, set_reference=True)
+
+        if type(self.postprocessing)==str:
+            self.postprocessing = dill.loads(self.postprocessing)
+        return graph
+
+
+    def _sample(self, graph):
```

Comment:
try to simplify _sample
it should just be:
while possible
  propose
  accept